File: common/common-ts/preflight/h5wasm-check.mjs
#!/usr/bin/env node
// Phase 0 Block 7 deliverable 0: h5wasm round-trip preflight.
//
// Writes a tiny HDF5 file with a single 4-element float dataset at
// /test/data; the companion h5wasm-check.py reads it via h5py and
// confirms the values round-trip. If h5wasm cannot install, build, or
// round-trip, the block must REFUTE and surface to the user.

import { existsSync, mkdirSync, rmSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import h5wasm from "h5wasm/node";

const here = dirname(fileURLToPath(import.meta.url));
const outDir = resolve(here, "out");
const outFile = resolve(outDir, "preflight.h5");

if (existsSync(outFile)) {
  rmSync(outFile);
}
mkdirSync(outDir, { recursive: true });

await h5wasm.ready;
const { File } = h5wasm;

const file = new File(outFile, "w");
try {
  const data = new Float64Array([1.5, -2.25, 3.125, 0.0]);
  file.create_group("test");
  file.get("test").create_dataset({
    name: "data",
    data,
    shape: [data.length],
    // h5wasm dtype codes: `<d` = float64, `<f` = float32 (h5wasm
    // README.md lines 254 + 263). The `<f8` form numpy uses is not
    // recognized by h5wasm and silently maps to float32.
    dtype: "<d",
  });
} finally {
  file.flush();
  file.close();
}

console.log(`wrote ${outFile}`);
console.log("values: [1.5, -2.25, 3.125, 0.0]");

File: common/common-ts/preflight/h5wasm-check.py
"""Phase 0 Block 7 deliverable 0 — Python half of the h5wasm round-trip.

Reads ``preflight/out/preflight.h5`` (written by ``h5wasm-check.mjs``)
via h5py and asserts the values match exactly.
"""

from __future__ import annotations

import sys
from pathlib import Path

import h5py
import numpy as np

HERE = Path(__file__).resolve().parent
H5_PATH = HERE / "out" / "preflight.h5"
EXPECTED = np.array([1.5, -2.25, 3.125, 0.0], dtype=np.float64)


def main() -> int:
    if not H5_PATH.exists():
        print(f"preflight HDF5 file not found at {H5_PATH}", file=sys.stderr)
        print("did you run `node h5wasm-check.mjs` first?", file=sys.stderr)
        return 1
    with h5py.File(H5_PATH, "r") as h:
        if "test/data" not in h:
            print("missing dataset /test/data", file=sys.stderr)
            return 1
        actual = np.asarray(h["test/data"][()])
    if actual.dtype != EXPECTED.dtype:
        print(
            f"dtype mismatch: expected {EXPECTED.dtype}, got {actual.dtype}",
            file=sys.stderr,
        )
        return 1
    if actual.shape != EXPECTED.shape:
        print(
            f"shape mismatch: expected {EXPECTED.shape}, got {actual.shape}",
            file=sys.stderr,
        )
        return 1
    if not np.array_equal(actual, EXPECTED):
        print(
            f"value mismatch: expected {EXPECTED.tolist()}, got {actual.tolist()}",
            file=sys.stderr,
        )
        return 1
    print(f"read {H5_PATH}: values = {actual.tolist()} — match")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
